### ISSUE TITLE
SWMR seem to be broken

### ISSUE THREAD

**Issue by semohr** [usnistgov/h5wasm#30]
Hi, I tried to get the swmr mode working for some time now but made no progress so far. Thus I thought you might have an idea...

**Setup:**
Overall, I'm trying to read a file with h5wasm while writing it with h5py. As both libraries wrap the same c code, this should theoretically be possible. I can read the file with h5wasm without any problem if I don't use swmr.

**Reader:** (h5wasm)
```javascript
import h5wasm from "h5wasm";

const { FS } = await h5wasm.ready;
var file = new h5wasm.File('./data.hdf5', "Sr");
```
**Error:**
The reader errors out with
```
HDF5-DIAG: Error detected in HDF5 (1.12.1) thread 0:
  #000: /home/brian/dev/libhdf5-wasm/wasm_build/1_12_1/_deps/hdf5-src/src/H5F.c line 493 in H5Fcreate(): invalid flags
    major: Invalid arguments to routine
    minor: Bad value
```

**Writer:** (h5py)
Only using h5py, swmr works flawlessly with multiple readers and a single writer
```python
import h5py
import time
f = h5py.File("data.hdf5", "w", libver="v110")
f.swmr_mode = True

f.create_dataset("nodes", (0, 4), maxshape=(None, 4), dtype=h5py.string_dtype())
while True:
   # write data
    s = ds.shape
    ds.resize((s[0] + 1, *s[1:]))
    ds[s[0]] = ["test0","test1","test2","test3"]
    ds.flush()

    # Wait 5 seconds
    time.sleep(5)
```

**Comment by bmaranville** [comment]
Thanks for the bug report - I had never tried this setup before (reading from h5wasm while writing with swmr).  I will reproduce using your very helpful notes and see what I find.

**Comment by bmaranville** [comment]
Ok, I think I have something working in v0.4.5 just released.  I had to fix a bug in the hdf5.File constructor (was using H5Fcreate for  mode 'Sr' but should have been using H5Fopen).
I also had to implement a 'refresh' operation.
I added a property ```auto_refresh``` to the Dataset class, which defaults to false.  If you are reading in SWMR mode and want to auto-refresh a dataset on each read, just ```ds.auto_refresh = true;``` after getting a handle to the dataset. The implementation actually triggers the H5Drefresh when the metadata is read, because

- the size is read from the metadata before getting the data in order to allocate memory
- the metadata is read exactly once before the data is retrieved, so that the shapes etc. should be consistent.

Example:
```js
import h5wasm from "h5wasm";

const { FS } = await h5wasm.ready;
var file = new h5wasm.File('./data.hdf5', "Sr");
let ds = file.get("nodes");
ds.auto_refresh = true;
ds.shape;
// returns 12
// ... later:
ds.shape;
// returns 16
```

**Comment by semohr** [comment]
Thank you very much I appreciate your work! That was quick :+1: Maybe you want to add the `auto_refresh` property to the readme somewhere, such that other people can find it as well.

**Comment by semohr** [comment]
Just tested it and it looks like the `ds.shape` getter is working but I can't get the values, nor the json array from the dataset.

The following only occurs **while** the writer is running:
```javascript
import h5wasm from "h5wasm";

const { FS } = await h5wasm.ready;
var file = new h5wasm.File('./data.hdf5', "Sr");
let ds = file.get("nodes");
ds.auto_refresh = true;
ds.shape;
// returns 12
// ... later:
ds.value
// throws error
ds.to_array()
// throws error
```
```
error - RuntimeError: memory access out of bounds

HDF5-DIAG: Error detected in HDF5 (1.12.1) thread 0:
  #000: /home/brian/dev/libhdf5-wasm/wasm_build/1_12_1/_deps/hdf5-src/src/H5Dio.c line 179 in H5Dread(): can't read data
    major: Dataset
    minor: Read failed
  #001: /home/brian/dev/libhdf5-wasm/wasm_build/1_12_1/_deps/hdf5-src/src/H5VLcallback.c line 2011 in H5VL_dataset_read(): dataset read failed
    major: Virtual Object Layer
    minor: Read failed
  #002: /home/brian/dev/libhdf5-wasm/wasm_build/1_12_1/_deps/hdf5-src/src/H5VLcallback.c line 1978 in H5VL__dataset_read(): dataset read failed
    major: Virtual Object Layer
    minor: Read failed
  #003: /home/brian/dev/libhdf5-wasm/wasm_build/1_12_1/_deps/hdf5-src/src/H5VLnative_dataset.c line 166 in H5VL__native_dataset_read(): can't read data
    major: Dataset
    minor: Read failed
  #004: /home/brian/dev/libhdf5-wasm/wasm_build/1_12_1/_deps/hdf5-src/src/H5Dio.c line 545 in H5D__read(): can't read data
    major: Dataset
    minor: Read failed
  #005: /home/brian/dev/libhdf5-wasm/wasm_build/1_12_1/_deps/hdf5-src/src/H5Dchunk.c line 2572 in H5D__chunk_read(): chunked read failed
    major: Dataset
    minor: Read failed
  #006: /home/brian/dev/libhdf5-wasm/wasm_build/1_12_1/_deps/hdf5-src/src/H5Dscatgath.c line 527 in H5D__scatgath_read(): datatype conversion failed
    major: Dataset
    minor: Can't convert datatypes
  #007: /home/brian/dev/libhdf5-wasm/wasm_build/1_12_1/_deps/hdf5-src/src/H5T.c line 5458 in H5T_convert(): datatype conversion failed
    major: Datatype
    minor: Can't convert datatypes
  #008: /home/brian/dev/libhdf5-wasm/wasm_build/1_12_1/_deps/hdf5-src/src/H5Tconv.c line 3344 in H5T__conv_vlen(): can't read VL data
    major: Datatype
    minor: Read failed
  #009: /home/brian/dev/libhdf5-wasm/wasm_build/1_12_1/_deps/hdf5-src/src/H5Tvlen.c line 936 in H5T__vlen_disk_read(): unable to get blob
    major: Datatype
    minor: Can't get value
  #010: /home/brian/dev/libhdf5-wasm/wasm_build/1_12_1/_deps/hdf5-src/src/H5VLcallback.c line 6918 in H5VL_blob_get(): blob get failed
    major: Virtual Object Layer
    minor: Can't get value
  #011: /home/brian/dev/libhdf5-wasm/wasm_build/1_12_1/_deps/hdf5-src/src/H5VLcallback.c line 6883 in H5VL__blob_get(): blob get callback failed
    major: Virtual Object Layer
    minor: Can't get value
  #012: /home/brian/dev/libhdf5-wasm/wasm_build/1_12_1/_deps/hdf5-src/src/H5VLnative_blob.c line 125 in H5VL__native_blob_get(): unable to read VL information
    major: Virtual Object Layer
    minor: Read failed
  #013: /home/brian/dev/libhdf5-wasm/wasm_build/1_12_1/_deps/hdf5-src/src/H5HG.c line 599 in H5HG_read(): unable to protect global heap
    major: Heap
    minor: Unable to protect metadata
  #014: /home/brian/dev/libhdf5-wasm/wasm_build/1_12_1/_deps/hdf5-src/src/H5HG.c line 256 in H5HG__protect(): unable to protect global heap
    major: Heap
    minor: Unable to protect metadata
  #015: /home/brian/dev/libhdf5-wasm/wasm_build/1_12_1/_deps/hdf5-src/src/H5AC.c line 1426 in H5AC_protect(): H5C_protect() failed
    major: Object cache
    minor: Unable to protect metadata
  #016: /home/brian/dev/libhdf5-wasm/wasm_build/1_12_1/_deps/hdf5-src/src/H5C.c line 2368 in H5C_protect(): can't load entry
    major: Object cache
    minor: Unable to load metadata into cache
  #017: /home/brian/dev/libhdf5-wasm/wasm_build/1_12_1/_deps/hdf5-src/src/H5C.c line 7137 in H5C__load_entry(): invalid len with respect to EOA
    major: Object cache
    minor: Bad value
  #018: /home/brian/dev/libhdf5-wasm/wasm_build/1_12_1/_deps/hdf5-src/src/H5C.c line 7054 in H5C__verify_len_eoa(): address of object past end of allocation
    major: Object cache
    minor: Bad value
Invalid UTF-8 leading byte 0x86 encountered when deserializing a UTF-8 string in wasm memory to a JS string!
```

**Comment by semohr** [comment]
I think I found the problem, one needs to not only flush the dataset (`ds.flush()`) but also the file (`f.flush()`). Otherwise the dataset is not written to disk but to a temp storage... Well this is very unintuitive.

For future reference:
```python
import h5py
import time
f = h5py.File("data.hdf5", "w", libver="v110")
f.swmr_mode = True

f.create_dataset("nodes", (0, 4), maxshape=(None, 4), dtype=h5py.string_dtype())
while True:
   # write data
    s = ds.shape
    ds.resize((s[0] + 1, *s[1:]))
    ds[s[0]] = ["test0","test1","test2","test3"]
    ds.flush()
    f.flush() # <-------- Important

    # Wait 5 seconds
    time.sleep(5)
```

**Comment by bmaranville** [comment]
Ah - I see.  I tested ds.value but not ds.to_array.  I was able to read with ds.value many times as the file was written without problems, but I can see in the h5wasm code that there will be an issue with to_array, as it accesses metadata directly as well as the data, which will sometimes cause the shape to be different between the metadata and the data (refresh gets called twice, should be only once).

**Comment by bmaranville** [comment]
Ok, this should be fixed in v0.4.6 - there is no auto_refresh (would have been nice, but caused too many problems) and instead you should call ```ds.refresh()``` every time you want to get the most recent dataset/metadata (including the ```to_array()``` method)
